testsuite/tests/typing-poly/poly_params.ml

```diff
@@ -0,0 +1,202 @@
+[%%expect {|
+- : int * int option = (8, Some 8)
+|}];;
+
+let _ = poly3 (fun x -> x + 1) 8
+[%%expect {|
+Line 1, characters 14-30:
+1 | let _ = poly3 (fun x -> x + 1) 8
+                  ^^^^^^^^^^^^^^^^
+Error: This argument has type int -> int which is less general than
+         'a. 'a -> 'a
+|}];;
+
+let rec poly4 p (id : 'a. 'a -> 'a) =
+  if p then poly4 false id else id 4, id "four"
+[%%expect {|
+val poly4 : bool -> ('a. 'a -> 'a) -> int * string = <fun>
+|}];;
+
+let _ = poly4 true (fun x -> x)
+[%%expect {|
+- : int * string = (4, "four")
+|}];;
+
+let _ = poly4 true (fun x -> x + 1)
+[%%expect {|
+Line 1, characters 19-35:
+1 | let _ = poly4 true (fun x -> x + 1)
+                       ^^^^^^^^^^^^^^^^
+Error: This argument has type int -> int which is less general than
+         'a. 'a -> 'a
+|}];;
+
+let rec poly5 : bool -> ('a. 'a -> 'a) -> int * string =
+  fun p id ->
+    if p then poly5 false id else id 5, id "five"
+[%%expect {|
+val poly5 : bool -> ('a. 'a -> 'a) -> int * string = <fun>
+|}];;
+
+let _ = poly5 true (fun x -> x)
+[%%expect {|
+- : int * string = (5, "five")
+|}];;
+
+let _ = poly5 true (fun x -> x + 1)
+[%%expect {|
+Line 1, characters 19-35:
+1 | let _ = poly5 true (fun x -> x + 1)
+                       ^^^^^^^^^^^^^^^^
+Error: This argument has type int -> int which is less general than
+         'a. 'a -> 'a
+|}];;
+
+
+let rec poly6 : 'b. bool -> ('a. 'a -> 'a) -> 'b -> 'b * 'b option =
+  fun p id x ->
+    if p then poly6 false id x else id x, id (Some x)
+[%%expect {|
+val poly6 : bool -> ('a. 'a -> 'a) -> 'b -> 'b * 'b option = <fun>
+|}];;
+
+let _ = poly6 true (fun x -> x) 8
+[%%expect {|
+- : int * int option = (8, Some 8)
+|}];;
+
+let _ = poly6 true (fun x -> x + 1) 8
+[%%expect {|
+Line 1, characters 19-35:
+1 | let _ = poly6 true (fun x -> x + 1) 8
+                       ^^^^^^^^^^^^^^^^
+Error: This argument has type int -> int which is less general than
+         'a. 'a -> 'a
+|}];;
+
+let needs_magic (magic : 'a 'b. 'a -> 'b) = (magic 5 : string)
+let _ = needs_magic (fun x -> x)
+[%%expect {|
+val needs_magic : ('a 'b. 'a -> 'b) -> string = <fun>
+Line 2, characters 20-32:
+2 | let _ = needs_magic (fun x -> x)
+                        ^^^^^^^^^^^^
+Error: This argument has type 'c. 'c -> 'c which is less general than
+         'a 'b. 'a -> 'b
+|}];;
+
+let with_id (f : ('a. 'a -> 'a) -> 'b) = f (fun x -> x)
+[%%expect {|
+val with_id : (('a. 'a -> 'a) -> 'b) -> 'b = <fun>
+|}];;
+
+let _ = with_id (fun id -> id 4, id "four")
+[%%expect {|
+- : int * string = (4, "four")
+|}];;
+
+let non_principal1 p f =
+  if p then with_id f
+  else f (fun x -> x)
+[%%expect {|
+val non_principal1 : bool -> (('a. 'a -> 'a) -> 'b) -> 'b = <fun>
+|}, Principal{|
+Line 3, characters 7-21:
+3 |   else f (fun x -> x)
+           ^^^^^^^^^^^^^^
+Warning 18 [not-principal]: applying a higher-rank function here is not principal.
+val non_principal1 : bool -> (('a. 'a -> 'a) -> 'b) -> 'b = <fun>
+|}];;
```

Comment:
It's not principal because you only happen to know that the `f` function has a polymorphic parameter because you typed the `then` part of the `if` before the `else` part. If you swap the conditions this will no longer type-check.